lib/chainweb/request.ex

```diff
@@ -0,0 +1,161 @@
+  @type encoded_query :: String.t() | nil
+  @type endpoint :: String.t() | nil
+  @type path :: String.t() | nil
+  @type segment :: String.t() | nil
+  @type opts :: Keyword.t()
+  @type params :: Keyword.t()
+  @type response :: {:ok, map()} | {:error, Error.t()}
+  @type parsed_response :: {:ok, struct()} | {:error, Error.t()}
+
+  @type t :: %__MODULE__{
+          method: method(),
+          api_type: api_type(),
+          network_id: network_id(),
+          chain_id: chain_id(),
+          endpoint: endpoint(),
+          path: path(),
+          segment: segment(),
+          segment_path: path(),
+          query: query(),
+          headers: headers(),
+          body: body(),
+          encoded_query: encoded_query()
+        }
+
+  defstruct [
+    :method,
+    :api_type,
+    :network_id,
+    :chain_id,
+    :endpoint,
+    :path,
+    :segment,
+    :segment_path,
+    :query,
+    :headers,
+    :body,
+    :encoded_query
+  ]
+
+  @spec new(method :: method(), opts :: opts()) :: t()
+  def new(method, [{api_type, opts}]) when api_type in [:pact, :p2p] do
+    network_id = Keyword.get(opts, :network_id)
+    chain_id = Keyword.get(opts, :chain_id)
+    endpoint = Keyword.get(opts, :endpoint)
+    path = Keyword.get(opts, :path)
+    segment = Keyword.get(opts, :segment)
+    segment_path = Keyword.get(opts, :segment_path)
+
+    %__MODULE__{
+      method: method,
+      api_type: api_type,
+      network_id: network_id,
+      chain_id: chain_id,
+      endpoint: endpoint,
+      path: path,
+      segment: segment,
+      segment_path: segment_path,
+      query: [],
+      headers: []
+    }
+  end
+
+  @spec set_chain_id(t(), chain_id :: chain_id()) :: t()
+  def set_chain_id(%__MODULE__{} = request, chain_id) do
+    %{request | chain_id: chain_id}
+  end
+
+  @spec set_network(t(), network :: network_id()) :: t()
+  def set_network(%__MODULE__{} = request, network) do
+    %{request | network_id: network}
+  end
+
+  @spec add_body(request :: t(), body :: body()) :: t()
+  def add_body(%__MODULE__{} = request, body) do
+    %{request | body: body}
+  end
+
+  @spec add_headers(request :: t(), headers :: headers()) :: t()
+  def add_headers(%__MODULE__{} = request, headers) do
+    %{request | headers: headers}
+  end
+
+  @spec add_query(request :: t(), params :: params()) :: t()
+  def add_query(%__MODULE__{} = request, params) do
+    %{request | query: params, encoded_query: build_query_string(params)}
+  end
+
+  @spec perform(request :: t()) :: response()
+  def perform(
+        %__MODULE__{
+          method: method,
+          headers: headers,
+          body: body
+        } = request
+      ) do
+    request
+    |> build_request_url()
+    |> (&Client.request(method, &1, headers, body)).()
+  end
+
+  @spec results(response :: response(), opts :: opts()) :: parsed_response()
+  def results({:ok, results}, as: resource) do
+    {:ok, results |> MapCase.to_snake!() |> resource.new()}
```

Comment:
🤔 What do you think about changing it for something like this? 
```suggestion
    resource = 
         results 
         |> MapCase.to_snake!() 
         |> resource.new()
    {:ok, resource}
```